### PR TITLE
Make fire leave burnt decals on the tiles

### LINKED ISSUE #28303: Fire should make tiles look burnt visually
## Description
Tile fires (ie. plasmafires) should make any tile they're on look burnt visually, like in SS13. This both visually looks very good and serves as an indicator for rooms that have been ravenged by a deadly plasmafire.

This is probably best accomplished using decals, since there are probably too many tiles to account for to have them transmute in to specific kinds (not to even mention stuff that isn't even "tiles" like grass.) Beestation has some icons for this in [icons/turf/turf_damage.dmi](https://scrubby.melonmesa.com/dmi/icons/turf/turf_damage.dmi?codebase=beestation).

### PR DESCRIPTION
**Pull request by Winkarst-cpu**
## About the PR
Made fire leave burnt decals on the tiles.

## Why / Balance
Resolves #28303.

## Technical details
Code has comments.

## Media

https://github.com/user-attachments/assets/f0b4b06d-72dd-43b5-8245-66a178592724

## Requirements

- [x] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- add: Now fire leaves burn marks on the tiles that were affected by it.

### CODE REVIEW

**Comment by Djungelskog2** [comment]
thank the lord this is finally a thing

**Comment by SoulFN** [comment]
Can you clean it?

**Comment by Winkarst-cpu** [comment]
> Can you clean it?

Yeah, with a space cleaner.

**Comment by Winkarst-cpu** [comment]
Unrelated failed check?...

**Comment by slarticodefast** [comment]
Might be related. I will re-run them to be sure.

**Comment by Winkarst-cpu** [comment]
@slarticodefast, it's fine now.

**Comment by Emisse** [comment]
seems fine idk about enumerating proto tags in the system but seems fine

File: Content.Server/Atmos/EntitySystems/AtmosphereSystem.Hotspot.cs [hunk 1]
```diff
@@ -56,7 +61,19 @@ private void ProcessHotspot(
...
+
+                // Get prototypes of the burnt tile decals
+                var decals = _prototypeManager.EnumeratePrototypes<DecalPrototype>().Where(x => x.Tags.Contains("burnt")).Select(x => x.ID);
+
+                // Get already existing burnt decals on the tile
+                var tileDecals = _decalSystem.GetDecalsInRange(gridUid, tilePos).Select(x => decals.Contains(x.Decal.Id));
+
+                // Add a random burned decal to the tile only if there are less than 4 of them
+                if (tileDecals.Count() < 4)
+                    _decalSystem.TryAddDecal(decals.ElementAt(_random.Next(decals.Count())), new(gridUid, tilePos), out _, cleanable: true);
```
Comment (metalgearsloth) [inline Content.Server/Atmos/EntitySystems/AtmosphereSystem.Hotspot.cs:76]: This is too slow to use at runtime:
1. Do not use LINQ in atmos code, it is too slow.
2. Do not iterate every single prototype, dictionaries are too slow.
3. .Count() is also LINQ.

**Review by metalgearsloth — DISMISSED** [review]

File: Content.Server/Atmos/EntitySystems/AtmosphereSystem.cs [hunk 2]
```diff
@@ -65,8 +71,10 @@ public override void Initialize()
         _airtightQuery = GetEntityQuery<AirtightComponent>();
         _firelockQuery = GetEntityQuery<FirelockComponent>();
 
-        SubscribeLocalEvent<TileChangedEvent>(OnTileChanged);
+        // Store burnt tiles decals for the future use
+        _burntDecals = _prototypeManager.EnumeratePrototypes<DecalPrototype>().Where(x => x.Tags.Contains("burnt")).Select(x => x.ID).ToList();
```
Comment (metalgearsloth) [inline Content.Server/Atmos/EntitySystems/AtmosphereSystem.cs:75]: Doesn't work with prototype reloads, just add a prototypemanager sub and refresh it there.

**Comment by slarticodefast** [comment]
Will all the burn decals cause performance problems after large plasma fires?

**Comment by Winkarst-cpu** [comment]
@slarticodefast, seems fine to me.


https://github.com/user-attachments/assets/57efbbef-70fc-4d09-8aa7-d0cf5bf38d96

**Comment by metalgearsloth** [comment]
> @slarticodefast, seems fine to me.
> 2024-09-14.23-48-51.mp4

They mean server mate.

File: Content.Server/Atmos/EntitySystems/AtmosphereSystem.cs [hunk 3]
```diff
@@ -107,4 +121,9 @@ public override void Update(float frameTime)
 
         _exposedTimer -= ExposedUpdateDelay;
     }
+
+    private void CacheDecals()
+    {
+        _burntDecals = _prototypeManager.EnumeratePrototypes<DecalPrototype>().Where(x => x.Tags.Contains("burnt")).Select(x => x.ID).ToList();
+    }
 }
```
Comment (metalgearsloth) [inline Content.Server/Atmos/EntitySystems/AtmosphereSystem.cs:129]: ToArray because you're never actually modifying this and they're faster than lists.

**Comment by metalgearsloth** [comment]
It's hard to profile atmos because it gets time-sliced but uhh no one really looks at its perf a lot so.

**Pull request merged**